src/genetic_algorithm/algorithms/species_based_algorithm.rs

```diff
@@ -0,0 +1,176 @@
+                    .zip(population.iter())
+                    .take(species.len() / 2)
+                {
+                    let (mut first_child, mut second_child) = first_parent.breed(&second_parent);
+
+                    first_child = first_child.mutate(1.0);
+                    second_child = second_child.mutate(1.0);
+                    first_child.correct();
+                    second_child.correct();
+                    offspring_population.push(first_child);
+                    offspring_population.push(second_child);
+                }
+            } else {
+                // Breed the species
+                for parents in species.chunks(2) {
+                    if parents.len() < 2 {
+                        continue;
+                    }
+
+                    let first_parent = &parents[0];
+                    let second_parent = &parents[1];
+
+                    let (mut first_child, mut second_child) = first_parent.breed(&second_parent);
+
+                    first_child = first_child.mutate(1.0);
+                    second_child = second_child.mutate(1.0);
+                    first_child.correct();
+                    second_child.correct();
+                    offspring_population.push(first_child);
+                    offspring_population.push(second_child);
+                }
+            }
+
+            if species.len() % 2 != 0 {
+                let first_parent = &species.last().unwrap();
+                let second_parent = population[species.len() - 1];
+                let (mut first_child, mut second_child) = first_parent.breed(&second_parent);
+
+                first_child = first_child.mutate(1.0);
+                second_child = second_child.mutate(1.0);
+                first_child.correct();
+                second_child.correct();
+                offspring_population.push(first_child);
+                offspring_population.push(second_child);
+            }
+        }
+
+        offspring_population.truncate(self.population.values().flatten().count());
+        self.offspring_population = offspring_population;
+
+        dbg!(&self.offspring_population.len());
+        dbg!(&self.population.values().flatten().count());
+    }
+
+    fn replacement(&mut self) {
+        self.population.clear();
+        for chromosome in self.new_population.iter() {
+            self.population
+                .entry(chromosome.characteristic())
+                .or_insert(Vec::new())
+                .push(chromosome.clone());
+        }
+        self.new_population.clear();
+    }
+
+    fn finished_evaluating(&mut self, chromosome: T) {
+        self.new_population.push(chromosome);
+    }
+
+    fn all_have_finished_evaluating(&self, population_size: usize) -> bool {
+        self.new_population.len() == population_size * 2
+    }
+
+    fn save_results(&self, generation_count: usize) {
+        let mut population: Vec<&T> = self.population.values().flatten().clone().collect();
```

Comment:
If you `clone`, why is it a `Vec<&T>` instead of a `Vec<T>`? Or does `values` return references? If it does, maybe the `clone` isn't needed

Reply:
Values return a vector of references. But we need to clone the vector because we will reorder it in a few lines later

Clone is needed